src/transformStateWithClones.js

```diff
@@ -7,7 +7,35 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const workingState = { ...state };
+  const stateHistory = [];
+
+  for (const action of actions) {
+    // Clear workingState
+    if (action.type === 'clear') {
+      for (const line in workingState) {
+        delete workingState[line];
+      }
+    }
+
+    // ADD property to the workingState
+    if (action.type === 'addProperties') {
+      for (const newProperty in action.extraData) {
+        workingState[newProperty] = action.extraData[newProperty];
+      }
+    }
+
+    // Remove property from the workingState
+    if (action.type === 'removeProperties') {
+      for (const removingKey of action.keysToRemove) {
+        delete workingState[removingKey];
+      }
+    }
+
+    stateHistory.push({ ...workingState });
```

Comment:
[DONT REPEAT YOURSELF] - You are pushing a copy of the object to the array at the end of each loop cycle which is good. However, this action is outside of the `if` blocks. To adhere to the DRY principle, consider restructuring your code to use a switch statement and perform this action once after the switch block.